src/main.cc

```diff
@@ -1443,12 +1434,57 @@ ConfigureCurrentKid(const CommandLine &cmdLine)
     }
 }
 
-static void StartUsingConfig()
+/// Start directing debugs() messages to the configured cache.log.
+/// Until this function is called, all allowed messages go to stderr.
+static void
+ConfigureDebugging()
+{
+    if (opt_no_daemon) {
+        fd_open(0, FD_LOG, "stdin");
+        fd_open(1, FD_LOG, "stdout");
+        fd_open(2, FD_LOG, "stderr");
+    }
+    // we should not create cache.log outside chroot environment, if any
+    // XXX: With Config.chroot_dir set, SMP master process never calls db_init().
+    if (!Config.chroot_dir || Chrooted)
+        _db_init(Debug::cache_log, Debug::debugOptions);
+}
+
+static void
+RunConfigUsers()
 {
     RunRegisteredHere(RegisteredRunner::claimMemoryNeeds);
     RunRegisteredHere(RegisteredRunner::useConfig);
 }
 
+static void
+StartUsingConfig()
+{
+    setMaxFD();
+    fde::Init();
+    const auto skipCwdAdjusting = IamMasterProcess() && InDaemonMode();
+    if (skipCwdAdjusting) {
+        ConfigureDebugging();
+        RunConfigUsers();
+    } else if (Config.chroot_dir) {
+        RunConfigUsers();
+        enter_suid();
+        // TODO: don't we need to RunConfigUsers() in the configured
+        // chroot environment?
+        mainSetCwd();
+        leave_suid();
+        ConfigureDebugging();
+    } else {
+        ConfigureDebugging();
```

Comment:
I am a little suspicious that this is different from the other else clause. This might put the debug log in whatever home directory, /tmp or such the init is using.

It seems to me they should both contain the same sequence - and that sequence should set the CWD as early as possible.

Reply:
It is a known fact that this sequence looks suspicious, for several reasons. Eduard's claim, AFAICT, is that essentially the same suspicious sequence is implemented in the official code. @yadij, do you have reasons to believe that the official sequence is different?

Please note that we should refrain from changing the existing sequence in _this_ PR (to the extent possible), regardless of whether that behavior is deemed right or wrong. The added TODOs already reflect known problems with the sequences. If you just want to add more TODOs, please say so explicitly.

> This might put the debug log in whatever home directory, /tmp or such the init is using.

Yes, it will do that in some cases, AFAICT. Eduard's claim, AFAICT, is that it would happen in the official code as well.
* If that claim is accurate, this PR can be merged without changes.
* If that claim is false, _and_ it is possible to preserve the official behavior without drastic changes, then this PR should be fixed to do so.
* If that claim is false but drastic changes would be required to preserve the official behavior, then the behavior change should be explained/documented. Depending on the details in that explanation/documentation, we would then decide whether this PR should be made conditional on a follow up PR that fixes the sequences (and changes the official behavior again) to be what we want them to be.

Reply:
> Amos (on [squid-dev](http://lists.squid-cache.org/pipermail/squid-dev/2019-October/009521.html)): I want to be very sure of the sequence being identical or better before
> this goes in.

@yadij, everybody wants that AFAICT, but it is not clear whether you are actively working on validating PR's "identical or better" claim, or you are just waiting for the PR author to do something that will assure you that those claims are correct. The PR is (or looks) stuck because of this uncertainty regarding your position. Please add a PR comment to resolve this uncertainty.

We should not rely on me _guessing_ that I should not clear this PR for merge until something happens on your side. This guessing and double checking wastes time and, sooner or later, I will guess incorrectly. AFAICT, we have more than enough mechanisms to prevent this guessing and associated stuck PRs. Please use them.

Reply:
I have now reviewed the call sequence and my suspicion was correct. One of these clause orderings is brand new code sequence. 

The order of RunConfigUsers() before ConfigureDebugging() is what the old code used to do. The first if-block and this final else block are new code. If that new sequence has been tested and works, then great - but we should keep the chroot sequence in sync.

 NP: doing the proposed shuffling of SUID bits into mainSetCwd() and then also moving this "if (Config.chroot_dir)" check to allow mainSetCwd to act as a no-op when irrelevant should resolve this confusing oddity. We would then have a much simpler:
``
 if (!skipCwdAdjusting) {
    mainSetCwd()
}
 ConfigureDebugging();
 RunConfigUsers();
``

PS. For completeness I found that the few code paths which skip or do not call this function at all are still doing what they used to. The technically unrelated move of setMaxFD and removal of TEST_ACCESS look fine.

Reply:
> The order of RunConfigUsers() before ConfigureDebugging() is what the old code used to do. 

To begin with, both of these methods are new: the old code could not use them. Basically, they correspond to the old StartUsingConfig() and _db_init() respectively, with the difference that ConfigureDebugging() does a little more, opening standard I/O streams. What do we see in the old code? When chroot is not configured, _db_init() is invoked always before StartUsingConfig() from parseConfigFile() internals:
```
ParseConfigFileOrThrow(const char *file_name) {
    [...]
    if (!Config.chroot_dir) {
        leave_suid();
        setUmask(Config.umask);
        _db_init(Debug::cache_log, Debug::debugOptions);
        enter_suid();
    }
}
```
This order is preserved in the StartUsingConfig().